File: app/routers/investment_transactions.py
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Account, Holding, InvestmentTransaction, LedgerTransaction
from app.schemas.account import (
    InvestmentTransactionCreate,
    InvestmentTransactionResponse,
    InvestmentTransactionUpdate,
)
from app.schemas.account import PaginatedResponse
from app.services.core import apply_buy, apply_sell, remove_holding, update_contributed_amount
from app.utils import (
    ASSET_CLASS_DEPOSIT,
    generate_deposit_symbol,
    holding_accrued_interest,
    is_deposit_holding,
    normalize_deposit_holding,
    to_decimal as dec,
)

router = APIRouter(prefix="/investment-transactions", tags=["investment-transactions"])


def _paginate(query, page: int, page_size: int):
    total = query.count()
    items = query.offset((page - 1) * page_size).limit(page_size).all()
    total_pages = (total + page_size - 1) // page_size if page_size else 1
    return items, total, total_pages


@router.get("", response_model=PaginatedResponse[InvestmentTransactionResponse])
def list_investment_transactions(
    account_id: int | None = None,
    holding_id: int | None = None,
    type: str | None = None,
    from_date: str | None = None,
    to_date: str | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=200),
    sort_order: str = "desc",
    db: Session = Depends(get_db),
):
    query = db.query(InvestmentTransaction)
    if account_id:
        query = query.filter(InvestmentTransaction.account_id == account_id)
    if holding_id:
        query = query.filter(InvestmentTransaction.holding_id == holding_id)
    if type:
        query = query.filter(InvestmentTransaction.type == type)
    if from_date:
        query = query.filter(InvestmentTransaction.transaction_date >= from_date)
    if to_date:
        query = query.filter(InvestmentTransaction.transaction_date <= to_date)
    if sort_order == "asc":
        query = query.order_by(InvestmentTransaction.transaction_date.asc(), InvestmentTransaction.id.asc())
    else:
        query = query.order_by(InvestmentTransaction.transaction_date.desc(), InvestmentTransaction.id.desc())
    items, total, total_pages = _paginate(query, page, page_size)
    return PaginatedResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages,
    )


@router.get("/{transaction_id}", response_model=InvestmentTransactionResponse)
def get_investment_transaction(transaction_id: int, db: Session = Depends(get_db)):
    tx = db.get(InvestmentTransaction, transaction_id)
    if not tx:
        raise HTTPException(status_code=404, detail="거래를 찾을 수 없습니다.")
    return tx


def _resolve_holding(db: Session, account: Account, payload: InvestmentTransactionCreate) -> Holding:
    if payload.holding_id:
        holding = db.get(Holding, payload.holding_id)
        if not holding or holding.account_id != account.id:
            raise HTTPException(status_code=400, detail="유효하지 않은 보유 종목입니다.")
        return holding

    asset_class = payload.asset_class or "stock"

    if asset_class == ASSET_CLASS_DEPOSIT:
        if not payload.name:
            raise HTTPException(status_code=400, detail="예금 상품명이 필요합니다.")
        holding = (
            db.query(Holding)
            .filter(
                Holding.account_id == account.id,
                Holding.asset_class == ASSET_CLASS_DEPOSIT,
                Holding.name == payload.name,
            )
            .first()
        )
        if not holding:
            holding = Holding(
                account_id=account.id,
                asset_class=ASSET_CLASS_DEPOSIT,
                symbol=payload.symbol or generate_deposit_symbol(),
                name=payload.name,
                quantity=0,
                avg_cost_price=Decimal("1"),
                manual_price=Decimal("1"),
            )
            db.add(holding)
            db.flush()
        return holding

    if not payload.symbol or not payload.name:
        raise HTTPException(status_code=400, detail="symbol과 name이 필요합니다.")
    holding = (
        db.query(Holding)
        .filter(Holding.account_id == account.id, Holding.symbol == payload.symbol)
        .first()
    )
    if not holding:
        holding = Holding(
            account_id=account.id,
            asset_class="stock",
            symbol=payload.symbol,
            name=payload.name,
            quantity=0,
            avg_cost_price=0,
        )
        db.add(holding)
        db.flush()
    return holding


def _apply_transaction_effects(db: Session, account: Account, holding: Holding | None, tx: InvestmentTransaction):
    amount = dec(tx.amount)
    fee = dec(tx.fee)
    if tx.type == "buy":
        if not holding:
            raise HTTPException(status_code=400, detail="매수 거래에는 보유 종목이 필요합니다.")
        if is_deposit_holding(holding):
            quantity = dec(tx.quantity) if tx.quantity else amount
            price = Decimal("1")
            normalize_deposit_holding(holding)
            apply_buy(holding, quantity, price)
        else:
            quantity = dec(tx.quantity)
            price = dec(tx.price)
            apply_buy(holding, quantity, price)
        account.cash_balance = dec(account.cash_balance) - amount - fee
    elif tx.type == "sell":
        if not holding:
            raise HTTPException(status_code=400, detail="매도 거래에는 보유 종목이 필요합니다.")
        sell_qty = dec(tx.quantity) if tx.quantity else amount
        apply_sell(holding, sell_qty)
        account.cash_balance = dec(account.cash_balance) + amount - fee
        if dec(holding.quantity) <= 0:
            tx.holding_id = None
            remove_holding(db, holding)
    elif tx.type == "deposit":
        account.cash_balance = dec(account.cash_balance) + amount
    elif tx.type in ("withdraw", "fee"):
        account.cash_balance = dec(account.cash_balance) - amount
    elif tx.type in ("dividend", "interest"):
        if holding and is_deposit_holding(holding) and tx.type == "interest":
            normalize_deposit_holding(holding)
            apply_buy(holding, amount, Decimal("1"))
            holding.start_date = tx.transaction_date
        else:
            account.cash_balance = dec(account.cash_balance) + amount
    update_contributed_amount(db, account.id, tx.transaction_date.year)


def _reverse_transaction_effects(db: Session, account: Account, tx: InvestmentTransaction):
    amount = dec(tx.amount)
    fee = dec(tx.fee)
    if tx.type == "buy":
        account.cash_balance = dec(account.cash_balance) + amount + fee
    elif tx.type == "sell":
        account.cash_balance = dec(account.cash_balance) - amount + fee
    elif tx.type == "deposit":
        account.cash_balance = dec(account.cash_balance) - amount
    elif tx.type in ("withdraw", "fee"):
        account.cash_balance = dec(account.cash_balance) + amount
    elif tx.type in ("dividend", "interest"):
        holding = db.get(Holding, tx.holding_id) if tx.holding_id else None
        if holding and is_deposit_holding(holding) and tx.type == "interest":
            apply_sell(holding, amount)
        else:
            account.cash_balance = dec(account.cash_balance) - amount


@router.post("", response_model=InvestmentTransactionResponse, status_code=201)
def create_investment_transaction(payload: InvestmentTransactionCreate, db: Session = Depends(get_db)):
    account = db.get(Account, payload.account_id)
    if not account:
        raise HTTPException(status_code=404, detail="계좌를 찾을 수 없습니다.")

    holding = None
    if payload.type in ("buy", "sell"):
        holding = _resolve_holding(db, account, payload)
    elif payload.holding_id:
        holding = db.get(Holding, payload.holding_id)
        if not holding or holding.account_id != account.id:
            raise HTTPException(status_code=400, detail="유효하지 않은 보유 종목입니다.")

    tx = InvestmentTransaction(
        account_id=payload.account_id,
        holding_id=holding.id if holding else None,
        type=payload.type,
        transaction_date=payload.transaction_date,
        quantity=payload.quantity,
        price=payload.price,
        amount=payload.amount,
        fee=payload.fee,
        memo=payload.memo,
    )
    db.add(tx)
    db.flush()
    _apply_transaction_effects(db, account, holding, tx)

    if payload.sync_to_ledger and payload.type in ("dividend", "interest"):
        if not payload.ledger_category_id:
            raise HTTPException(status_code=400, detail="가계부 연동 시 ledger_category_id가 필요합니다.")
        ledger_tx = LedgerTransaction(
            transaction_date=payload.transaction_date,
            type="income",
            amount=payload.amount,
            category_id=payload.ledger_category_id,
            merchant=holding.name if holding else account.name,
            memo=payload.memo or f"투자 {payload.type}",
        )
        db.add(ledger_tx)

    db.commit()
    db.refresh(tx)
    return tx


@router.patch("/{transaction_id}", response_model=InvestmentTransactionResponse)
def update_investment_transaction(
    transaction_id: int,
    payload: InvestmentTransactionUpdate,
    db: Session = Depends(get_db),
):
    tx = db.get(InvestmentTransaction, transaction_id)
    if not tx:
        raise HTTPException(status_code=404, detail="거래를 찾을 수 없습니다.")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(tx, key, value)
    db.commit()
    db.refresh(tx)
    update_contributed_amount(db, tx.account_id, tx.transaction_date.year)
    return tx


@router.delete("/{transaction_id}", status_code=204)
def delete_investment_transaction(transaction_id: int, db: Session = Depends(get_db)):
    tx = db.get(InvestmentTransaction, transaction_id)
    if not tx:
        raise HTTPException(status_code=404, detail="거래를 찾을 수 없습니다.")
    account = db.get(Account, tx.account_id)
    if not account:
        raise HTTPException(status_code=404, detail="계좌를 찾을 수 없습니다.")
    account_id = tx.account_id
    year = tx.transaction_date.year
    _reverse_transaction_effects(db, account, tx)
    db.delete(tx)
    db.commit()
    update_contributed_amount(db, account_id, year)

File: src/lib/services/investment-transactions.ts
import type { Account, AccountType, Holding, InvestmentTransaction, Prisma } from "@prisma/client";
import Decimal from "decimal.js";
import type { z } from "zod";

import { toDecimal } from "@/lib/decimal";
import { ServiceError } from "@/lib/service-error";
import {
  applyBuy,
  applySell,
  updateContributedAmount,
} from "@/lib/services/core";
import {
  adjustBookCostOnBuy,
  adjustBookCostOnSell,
  ASSET_CLASS_DEPOSIT,
  generateDepositSymbol,
  holdingMarketValue,
  isDepositHolding,
  normalizeDepositHolding,
  restoreBookCostOnSellReversal,
} from "@/lib/utils";
import type {
  investmentTransactionCreateSchema,
  investmentTransactionUpdateSchema,
} from "@/lib/validations/account";
import { prisma } from "@/lib/db";

type TxClient = Prisma.TransactionClient;
type InvestmentCreate = z.infer<typeof investmentTransactionCreateSchema>;
type InvestmentUpdate = z.infer<typeof investmentTransactionUpdateSchema>;

type AccountWithType = Account & { account_type: AccountType };

export function assertTransactionAllowedForAccount(
  account: AccountWithType,
  type: string,
): void {
  const { category, supports_holdings } = account.account_type;

  if (type === "buy" || type === "sell") {
    if (!supports_holdings) {
      throw new ServiceError(400, "이 계좌 유형은 매수·매도를 지원하지 않습니다.");
    }
    return;
  }

  if (type === "dividend") {
    if (!supports_holdings) {
      throw new ServiceError(400, "이 계좌 유형은 배당을 지원하지 않습니다.");
    }
    return;
  }

  if (type === "interest") {
    if (category !== "deposit" && !supports_holdings) {
      throw new ServiceError(
        400,
        "이자는 예적금 또는 보유 상품이 있는 계좌에서만 등록할 수 있습니다.",
      );
    }
    return;
  }
}

export async function resolveHolding(
  db: TxClient,
  account: Account,
  payload: InvestmentCreate,
): Promise<Holding> {
  if (payload.holding_id) {
    const holding = await db.holding.findUnique({ where: { id: payload.holding_id } });
    if (!holding || holding.account_id !== account.id) {
      throw new ServiceError(400, "유효하지 않은 보유 종목입니다.");
    }
    return holding;
  }

  const assetClass = payload.asset_class ?? "stock";

  if (assetClass === ASSET_CLASS_DEPOSIT) {
    if (!payload.name) {
      throw new ServiceError(400, "예금 상품명이 필요합니다.");
    }
    let holding = await db.holding.findFirst({
      where: {
        account_id: account.id,
        asset_class: ASSET_CLASS_DEPOSIT,
        name: payload.name,
      },
    });
    if (!holding) {
      holding = await db.holding.create({
        data: {
          account_id: account.id,
          asset_class: ASSET_CLASS_DEPOSIT,
          symbol: payload.symbol ?? generateDepositSymbol(),
          name: payload.name,
          quantity: 0,
          avg_cost_price: 1,
          manual_price: 1,
        },
      });
    }
    return holding;
  }

  if (!payload.symbol || !payload.name) {
    throw new ServiceError(400, "symbol과 name이 필요합니다.");
  }

  let holding = await db.holding.findFirst({
    where: { account_id: account.id, symbol: payload.symbol },
  });
  if (!holding) {
    holding = await db.holding.create({
      data: {
        account_id: account.id,
        asset_class: "stock",
        symbol: payload.symbol,
        name: payload.name,
        quantity: 0,
        avg_cost_price: 0,
      },
    });
  }
  return holding;
}

export async function applyTransactionEffects(
  db: TxClient,
  account: Account,
  holding: Holding | null,
  tx: InvestmentTransaction,
): Promise<void> {
  const amount = toDecimal(tx.amount);
  const fee = toDecimal(tx.fee);
  const tax = toDecimal(tx.tax);
  let cashBalance = toDecimal(account.cash_balance);
  let depositSellData: { amount: Decimal; quantity: Decimal } | null = null;

  switch (tx.type) {
    case "buy": {
      if (!holding) {
        throw new ServiceError(400, "매수 거래에는 보유 종목이 필요합니다.");
      }
      if (isDepositHolding(holding)) {
        const quantity = tx.quantity != null ? toDecimal(tx.quantity) : amount;
        normalizeDepositHolding(holding);
        applyBuy(holding, quantity, new Decimal(1));
      } else {
        applyBuy(holding, toDecimal(tx.quantity), toDecimal(tx.price), amount);
        adjustBookCostOnBuy(holding, amount);
      }
      cashBalance = cashBalance.minus(amount).minus(fee);
      break;
    }
    case "sell": {
      if (!holding) {
        throw new ServiceError(400, "매도 거래에는 보유 종목이 필요합니다.");
      }
      if (isDepositHolding(holding)) {
        const principal = toDecimal(holding.quantity);
        if (principal.lte(0)) {
          throw new ServiceError(400, "해지할 예금 원금이 없습니다.");
        }
        const redemptionAmount = holdingMarketValue(holding, tx.transaction_date);
        applySell(holding, principal);
        cashBalance = cashBalance.plus(redemptionAmount);
        depositSellData = { amount: redemptionAmount, quantity: principal };
        tx.amount = redemptionAmount;
        tx.quantity = principal;
        tx.price = new Decimal(1);
      } else {
        const sellQty = tx.quantity != null ? toDecimal(tx.quantity) : amount;
        const priorQty = toDecimal(holding.quantity);
        adjustBookCostOnSell(holding, sellQty, priorQty);
        applySell(holding, sellQty);
        cashBalance = cashBalance.plus(amount).minus(fee).minus(tax);
      }
      break;
    }
    case "deposit":
      cashBalance = cashBalance.plus(amount);
      break;
    case "withdraw":
    case "fee":
      cashBalance = cashBalance.minus(amount);
      break;
    case "dividend":
    case "interest": {
      if (holding && isDepositHolding(holding) && tx.type === "interest") {
        normalizeDepositHolding(holding);
        applyBuy(holding, amount, new Decimal(1));
        holding.start_date = tx.transaction_date;
      } else {
        cashBalance = cashBalance.plus(amount);
      }
      break;
    }
  }

  await db.account.update({
    where: { id: account.id },
    data: { cash_balance: cashBalance },
  });
  account.cash_balance = cashBalance;

  if (holding) {
    const exists = await db.holding.findUnique({ where: { id: holding.id } });
    if (exists) {
      await db.holding.update({
        where: { id: holding.id },
        data: {
          quantity: holding.quantity,
          avg_cost_price: holding.avg_cost_price,
          book_cost: holding.book_cost,
          manual_price: holding.manual_price,
          start_date: holding.start_date,
        },
      });
    }
  }

  if (tx.id && depositSellData) {
    await db.investmentTransaction.update({
      where: { id: tx.id },
      data: {
        amount: depositSellData.amount,
        quantity: depositSellData.quantity,
        price: new Decimal(1),
        fee: new Decimal(0),
        tax: new Decimal(0),
      },
    });
  }

  await updateContributedAmount(db, account.id, tx.transaction_date.getFullYear());
}

export async function reverseTransactionEffects(
  db: TxClient,
  account: Account,
  tx: InvestmentTransaction,
): Promise<void> {
  const amount = toDecimal(tx.amount);
  const fee = toDecimal(tx.fee);
  const tax = toDecimal(tx.tax);
  let cashBalance = toDecimal(account.cash_balance);

  switch (tx.type) {
    case "buy": {
      cashBalance = cashBalance.plus(amount).plus(fee);
      const holding = tx.holding_id
        ? await db.holding.findUnique({ where: { id: tx.holding_id } })
        : null;
      if (holding) {
        const sellQty = tx.quantity != null ? toDecimal(tx.quantity) : amount;
        if (!isDepositHolding(holding)) {
          holding.book_cost = toDecimal(holding.book_cost).minus(amount);
          if (toDecimal(holding.book_cost).lt(0)) {
            holding.book_cost = new Decimal(0);
          }
        }
        applySell(holding, sellQty);
        await db.holding.update({
          where: { id: holding.id },
          data: {
            quantity: holding.quantity,
            book_cost: holding.book_cost,
          },
        });
      }
      break;
    }
    case "sell": {
      cashBalance = cashBalance.minus(amount).plus(fee).plus(tax);
      const holding = tx.holding_id
        ? await db.holding.findUnique({ where: { id: tx.holding_id } })
        : null;
      if (holding) {
        const buyQty = tx.quantity != null ? toDecimal(tx.quantity) : amount;
        const buyPrice = tx.price != null ? toDecimal(tx.price) : toDecimal(holding.avg_cost_price);
        if (isDepositHolding(holding)) {
          applyBuy(holding, buyQty, new Decimal(1));
        } else {
          restoreBookCostOnSellReversal(holding, buyQty);
          applyBuy(holding, buyQty, buyPrice, amount);
        }
        await db.holding.update({
          where: { id: holding.id },
          data: {
            quantity: holding.quantity,
            avg_cost_price: holding.avg_cost_price,
            book_cost: holding.book_cost,
          },
        });
      }
      break;
    }
    case "deposit":
      cashBalance = cashBalance.minus(amount);
      break;
    case "withdraw":
    case "fee":
      cashBalance = cashBalance.plus(amount);
      break;
    case "dividend":
    case "interest": {
      const holding = tx.holding_id
        ? await db.holding.findUnique({ where: { id: tx.holding_id } })
        : null;
      if (holding && isDepositHolding(holding) && tx.type === "interest") {
        applySell(holding, amount);
        await db.holding.update({
          where: { id: holding.id },
          data: { quantity: holding.quantity },
        });
      } else {
        cashBalance = cashBalance.minus(amount);
      }
      break;
    }
  }

  await db.account.update({
    where: { id: account.id },
    data: { cash_balance: cashBalance },
  });
}

export async function updateInvestmentTransaction(
  transactionId: number,
  payload: InvestmentUpdate,
): Promise<InvestmentTransaction> {
  return prisma.$transaction(async (db) => {
    const existing = await db.investmentTransaction.findUnique({ where: { id: transactionId } });
    if (!existing) {
      throw new ServiceError(404, "거래를 찾을 수 없습니다.");
    }

    const account = await db.account.findUnique({
      where: { id: existing.account_id },
      include: { account_type: true },
    });
    if (!account) {
      throw new ServiceError(404, "계좌를 찾을 수 없습니다.");
    }

    const effectiveType = payload.type ?? existing.type;
    assertTransactionAllowedForAccount(account, effectiveType);

    const oldYear = existing.transaction_date.getFullYear();

    await reverseTransactionEffects(db, account, existing);

    const updated = await db.investmentTransaction.update({
      where: { id: transactionId },
      data: payload,
    });

    const holding = updated.holding_id
      ? await db.holding.findUnique({ where: { id: updated.holding_id } })
      : null;

    const refreshedAccount = await db.account.findUniqueOrThrow({ where: { id: account.id } });
    await applyTransactionEffects(db, refreshedAccount, holding, updated);

    const newYear = updated.transaction_date.getFullYear();
    await updateContributedAmount(db, account.id, oldYear);
    if (newYear !== oldYear) {
      await updateContributedAmount(db, account.id, newYear);
    }

    return updated;
  });
}
